src/core/ItemAtt.cpp

```diff
@@ -603,3 +603,118 @@ bool CItemAtt::Matches(time_t time1, time_t time2, int iObject,
     return PWSMatch::Match(time1, time2, testtime, iFunction);
   }
 }
+
+void CItemAtt::SerializePlainText(vector<char> &v)  const
+{
+  uuid_array_t uuid_array;
+  time_t t = 0;
+    
+  v.clear();
+  
+  // write mandatoty field ATTUUID
+  v.push_back(ATTUUID);
+  GetUUID(uuid_array);
+  CItem::push_length(v, sizeof(uuid_array_t));
+  v.insert(v.end(), uuid_array, (uuid_array + sizeof(uuid_array_t)));
+    
+  if(IsTitleSet())
+    push(v, ATTTITLE, GetTitle());
+    
+  if(IsCreationTimeSet()) {
+    GetCTime(t);
+    push(v, ATTCTIME, t);
+  }
+
+  if(IsFieldSet(MEDIATYPE))
+    push(v, MEDIATYPE, GetMediaType());
+  if(IsFieldSet(FILENAME))
+    push(v, FILENAME, GetFileName());
+  if(IsFieldSet(FILEPATH))
+    push(v, FILEPATH, GetFilePath());
+    
+  if(HasContent()) {
+    size_t length = GetContentSize();
+    size_t realLen = GetContentLength();
+    if(length) {
+      unsigned char data[length];
+      if(GetContent(data, length)) {
+        v.push_back(CONTENT);
+        push_length(v, static_cast<uint32>(realLen));
+        v.insert(v.end(), data, (data + realLen));
+      }
+    }
+  }
+    
+  if(IsFieldSet(FILECTIME)) {
+    GetFileCTime(t);
+    push(v, FILECTIME, t);
+  }
+  if(IsFieldSet(FILEMTIME)) {
+    GetFileMTime(t);
+    push(v, FILEMTIME, t);
+  }
+  if(IsFieldSet(FILEATIME)) {
+    GetFileATime(t);
+    push(v, FILEATIME, t);
+  }
+    
+  int end = END; // just to keep the compiler happy...
+  v.push_back(static_cast<char>(end));
+  push_length(v, 0);
+}
+
+
+bool CItemAtt::DeSerializePlainText(const std::vector<char> &v)
+{
+    auto iter = v.begin();
+    int emergencyExit = 255;
+    
+    while (iter != v.end()) {
+      unsigned char type = *iter++;
+      if (static_cast<uint32>(distance(v.end(), iter)) < sizeof(uint32)) {
+        ASSERT(0); // type must ALWAYS be followed by length
+        return false;
+      }
+      if (type == END) {
+        return true; // happy end
+      }
+        
+      uint32 len = *(reinterpret_cast<const uint32 *>(&(*iter)));
+      ASSERT(len < v.size()); // sanity check
+      iter += sizeof(uint32);
+
+      if (--emergencyExit == 0) {
```

Comment:
What does this variable do? Does it keep track of the number of fields read so far? If so, can we name it appropriately?

Reply:
The code is similar to the „old“ code in ItemData.cpp. The check is used to avoid an endless loop - the number of read values for an entry is limited, but in any case auf fault in memory, reading again and again the same place in memory due to error in RAM this loop is breaking the reading. As every check, in normal case that should never happen. It had been overtaken from original design in ItemData.cpp.